tests/psp721.tests.ts

```diff
@@ -0,0 +1,121 @@
+        const {
+            contract,
+            defaultSigner: sender,
+        } = await setup()
+
+        const ALICE = await getSigner('Alice')
+
+        // transfer from sender to Alice (token 0 was minted in mypsp_721 constructor)
+        await expect(() => contract.tx.transfer_from(sender.address, ALICE.address, 0)).to.changeTokenBalance(contract, ALICE, 1)
+        // allow sender to spend Alice's tokens
+        await fromSigner(contract, ALICE.address).tx.set_approval_for_all(sender.address, true)
+        // transfer from Alice back to sender
+        await expect(() => contract.tx.transfer_from(ALICE.address, sender.address, 0)).to.changeTokenBalance(contract, sender, 1)
+    })
+
+    it('Can not transfer non-existing token', async () => {
+        const {
+            contract,
+            accounts: [receiver],
+            defaultSigner: sender,
+        } = await setup()
+
+        await expect(contract.tx.transfer_from(sender.address, receiver.address, 1)).to.eventually.be.rejected
+    })
+
+    it('Can not transfer without allowance', async () => {
+        const {
+            contract,
+            accounts: [receiver],
+            defaultSigner: sender,
+        } = await setup()
+
+        const ALICE = await getSigner('Alice')
+        await expect(fromSigner(contract, ALICE.address).tx.transfer_from(sender.address, receiver.address, 0))
+            .to.eventually.be.rejected
+    })
+
+    it('Mint works', async () => {
+        const {
+            contract,
+            defaultSigner: sender,
+        } = await setup()
+
+        const ALICE = await getSigner('Alice')
+        await expect(() => contract.tx.mint(1)).to.changeTokenBalance(contract, sender, 1)
+        await expect(() => contract.tx.mint_to(ALICE.address, 2)).to.changeTokenBalance(contract, ALICE, 1)
+    })
+
+    it('Burn works', async () => {
+        const {
+            contract,
+            defaultSigner: sender,
+        } = await setup()
+
+        await expect(() => contract.tx.burn(0)).to.changeTokenBalance(contract, sender, -1)
+    })
+
+    it('Burn from works', async () => {
+        const {
+            contract,
+            defaultSigner: sender,
+        } = await setup()
+
+        const ALICE = await getSigner('Alice')
+
+        // allow ALICE to spend sender's tokens
+        await contract.tx.set_approval_for_all(ALICE.address, true)
+        // transfer from Alice back to sender
+        await expect(fromSigner(contract, ALICE.address).tx.burn_from(sender.address, 0))
+            .to.changeTokenBalance(contract, sender, -1)
+
+    })
+
+    it('Mint existing', async () => {
+        const {
+            contract,
+        } = await setup()
+
+        const ALICE = await getSigner('Alice')
+        await expect(() => contract.tx.mint(0)).to.eventually.be.rejected
+        await expect(() => contract.tx.mint_to(ALICE.address, 0)).to.eventually.be.rejected
+    })
+
+})
```

Comment:
Can you rearrange your code to match AAA pattern? Have a look at ownable.tests.ts